codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java

```diff
@@ -0,0 +1,843 @@
+
+    @Override
+    public Http2Headers authority(CharSequence value) {
+        throw new UnsupportedOperationException("read only");
+    }
+
+    @Override
+    public Http2Headers path(CharSequence value) {
+        throw new UnsupportedOperationException("read only");
+    }
+
+    @Override
+    public Http2Headers status(CharSequence value) {
+        throw new UnsupportedOperationException("read only");
+    }
+
+    @Override
+    public CharSequence method() {
+        return get(PseudoHeaderName.METHOD.value());
+    }
+
+    @Override
+    public CharSequence scheme() {
+        return get(PseudoHeaderName.SCHEME.value());
+    }
+
+    @Override
+    public CharSequence authority() {
+        return get(PseudoHeaderName.AUTHORITY.value());
+    }
+
+    @Override
+    public CharSequence path() {
+        return get(PseudoHeaderName.PATH.value());
+    }
+
+    @Override
+    public CharSequence status() {
+        return get(PseudoHeaderName.STATUS.value());
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder(getClass().getSimpleName()).append('[');
+        String separator = "";
+        for (Map.Entry<CharSequence, CharSequence> entry : this) {
+            builder.append(separator);
+            builder.append(entry.getKey()).append(": ").append(entry.getValue());
+            separator = ", ";
+        }
+        return builder.append(']').toString();
+    }
+
+    private final class ReadOnlyIterator implements Map.Entry<CharSequence, CharSequence>,
+                                                    Iterator<Map.Entry<CharSequence, CharSequence>> {
+        private int i;
+        private AsciiString[] current = pseudoHeaders.length != 0 ? pseudoHeaders : otherHeaders;
+        private AsciiString key;
+        private AsciiString value;
+
+        @Override
+        public boolean hasNext() {
+            return i != current.length;
+        }
+
+        @Override
+        public Map.Entry<CharSequence, CharSequence> next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            key = current[i];
+            value = current[++i];
+            if (++i == current.length && current == pseudoHeaders) {
+                current = otherHeaders;
+                i = 0;
+            }
+            return this;
+        }
+
+        @Override
+        public CharSequence getKey() {
+            return key;
+        }
+
+        @Override
+        public CharSequence getValue() {
+            return value;
+        }
+
+        @Override
+        public CharSequence setValue(CharSequence value) {
+            throw new UnsupportedOperationException("read only");
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException("read only");
+        }
+
+        @Override
+        public String toString() {
+            return key.toString() + '=' + value.toString();
```

Comment:
if we are concerned about the object state being mutated then perhaps we shouldn't have a single object represent  `Iterator` and `Entry`. However the `toString()` may still provide a useful snapshot into the object state for debugging etc. Are you concerned this may be misleading or miss-used in some way?